File: backend/app/schemas/interview.py
"""A3 动态访谈、结构化学生画像与确认门 Schema。"""

from __future__ import annotations

from enum import Enum
from typing import Any, Literal
from uuid import uuid4

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
    model_validator,
)


class InterviewStatus(str, Enum):
    IN_PROGRESS = "in_progress"
    AWAITING_CONFIRMATION = "awaiting_confirmation"
    CONFIRMED = "confirmed"


class InterviewDimension(str, Enum):
    RESEARCH_INTERESTS = "research_interests"
    RESEARCH_MODE = "research_mode"
    MENTORSHIP_STYLE = "mentorship_style"
    CAREER_ORIENTATION = "career_orientation"
    INNOVATION_RISK = "innovation_risk"
    HARD_CONSTRAINTS = "hard_constraints"


class HardConstraintField(str, Enum):
    LOCATION = "location"
    WEEKLY_COMMITMENT_DAYS = "weekly_commitment_days"
    DEGREE_STAGE = "degree_stage"
    LANGUAGE = "language"
    CONFIDENTIALITY = "confidentiality"
    GRADUATION_ARRANGEMENT = "graduation_arrangement"
    DEPARTMENT = "department"
    RESEARCH_TOPIC = "research_topic"
    ADVISOR_ID = "advisor_id"


class HardConstraintOperator(str, Enum):
    EQUALS = "equals"
    ONE_OF = "one_of"
    EXCLUDES = "excludes"
    CONTAINS = "contains"
    MINIMUM = "minimum"
    MAXIMUM = "maximum"


class HardConstraint(BaseModel):
    """由用户在画像编辑/确认阶段确认的结构化硬约束。"""

    model_config = ConfigDict(extra="forbid")

    field: HardConstraintField
    operator: HardConstraintOperator
    value: list[str] = Field(min_length=1, max_length=12)
    source_text: str | None = Field(default=None, max_length=500)

    @field_validator("value")
    @classmethod
    def normalize_values(cls, values: list[str]) -> list[str]:
        normalized: list[str] = []
        for value in values:
            item = value.strip()
            if item and item not in normalized:
                normalized.append(item)
        if not normalized:
            raise ValueError("硬约束 value 不得为空")
        return normalized

    @model_validator(mode="after")
    def validate_field_operator(self) -> "HardConstraint":
        if self.operator in {
            HardConstraintOperator.MINIMUM,
            HardConstraintOperator.MAXIMUM,
        }:
            if self.field != HardConstraintField.WEEKLY_COMMITMENT_DAYS:
                raise ValueError("minimum/maximum 仅适用于 weekly_commitment_days")
            if len(self.value) != 1:
                raise ValueError("数值硬约束只能提供一个值")
            try:
                number = float(self.value[0])
            except ValueError as exc:
                raise ValueError("每周投入天数必须是数值") from exc
            if not 0 <= number <= 7:
                raise ValueError("每周投入天数必须在 0 到 7 之间")
        return self


class DraftHardConstraint(BaseModel):
    """自然语言解析草案；用户确认前不得进入硬过滤。"""

    model_config = ConfigDict(extra="forbid")

    draft_id: str = Field(default_factory=lambda: str(uuid4()))
    source_text: str = Field(min_length=1, max_length=500)
    proposed_constraint: HardConstraint | None = None
    parsing_confidence: float = Field(ge=0, le=1)
    confirmation_prompt: str = Field(min_length=1, max_length=1000)


class StudentPortrait(BaseModel):
    """由学生回答或显式编辑得到的画像；不包含导师排序分数。"""

    model_config = ConfigDict(extra="forbid")

    research_interests: list[str] = Field(default_factory=list, max_length=8)
    interest_statement: str | None = Field(default=None, max_length=1000)
    research_mode: Literal["theory", "engineering", "mixed", "undecided"] | None = None
    mentorship_style: Literal[
        "high_guidance", "balanced", "autonomous", "undecided"
    ] | None = None
    career_orientation: Literal[
        "academic", "industry", "national_mission", "mixed", "undecided"
    ] | None = None
    innovation_risk: Literal["pioneering", "balanced", "mature", "undecided"] | None = None
    hard_constraints: list[HardConstraint] | None = Field(
        default=None, max_length=12
    )
    draft_hard_constraints: list[DraftHardConstraint] = Field(
        default_factory=list, max_length=12
    )
    unresolved_hard_constraints: list[str] | None = Field(
        default=None, max_length=12
    )

    @model_validator(mode="before")
    @classmethod
    def migrate_legacy_constraints(cls, value: Any) -> Any:
        """旧会话的自然语言约束只迁入 unresolved，不做语义猜测。"""
        if not isinstance(value, dict):
            return value
        raw_constraints = value.get("hard_constraints")
        if isinstance(raw_constraints, list) and any(
            isinstance(item, str) for item in raw_constraints
        ):
            migrated = dict(value)
            legacy = [
                item.strip()
                for item in raw_constraints
                if isinstance(item, str) and item.strip()
            ]
            existing = migrated.get("unresolved_hard_constraints") or []
            migrated["hard_constraints"] = []
            migrated["unresolved_hard_constraints"] = list(
                dict.fromkeys([*existing, *legacy])
            )
            return migrated
        return value

    @field_validator("research_interests")
    @classmethod
    def normalize_interests(cls, values: list[str]) -> list[str]:
        normalized: list[str] = []
        for value in values:
            item = value.strip()
            if item and item not in normalized:
                normalized.append(item)
        return normalized

    @field_validator("unresolved_hard_constraints")
    @classmethod
    def normalize_constraints(cls, values: list[str] | None) -> list[str] | None:
        if values is None:
            return None
        normalized: list[str] = []
        for value in values:
            item = value.strip()
            if item and item not in normalized:
                normalized.append(item)
        return normalized


class StudentPortraitPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)
    research_interests: list[str] | None = Field(default=None, max_length=8)
    interest_statement: str | None = Field(default=None, max_length=1000)
    research_mode: Literal["theory", "engineering", "mixed", "undecided"] | None = None
    mentorship_style: Literal[
        "high_guidance", "balanced", "autonomous", "undecided"
    ] | None = None
    career_orientation: Literal[
        "academic", "industry", "national_mission", "mixed", "undecided"
    ] | None = None
    innovation_risk: Literal["pioneering", "balanced", "mature", "undecided"] | None = None
    hard_constraints: list[HardConstraint] | None = Field(
        default=None, max_length=12
    )
    draft_hard_constraints: list[DraftHardConstraint] | None = Field(
        default=None, max_length=12
    )
    unresolved_hard_constraints: list[str] | None = Field(
        default=None, max_length=12
    )


class InterviewQuestionOption(BaseModel):
    model_config = ConfigDict(extra="forbid")

    value: str
    label: str


class InterviewQuestion(BaseModel):
    model_config = ConfigDict(extra="forbid")

    question_id: str
    dimension: InterviewDimension
    prompt: str
    answer_type: Literal["text", "single_choice"]
    options: list[InterviewQuestionOption] = Field(default_factory=list)
    information_goal: str


class InterviewMessage(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role: Literal["user", "assistant"]
    content: str
    created_at: str


class InterviewCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    initial_answer: str | None = Field(default=None, min_length=1, max_length=4000)


class InterviewAnswerRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    answer: str = Field(min_length=1, max_length=4000)


class InterviewConfirmRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)


class InterviewStateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    session_id: str
    status: InterviewStatus
    profile: StudentPortrait
    profile_version: int
    current_question: InterviewQuestion | None
    completed_dimensions: list[InterviewDimension]
    missing_dimensions: list[InterviewDimension]
    needs_confirmation: bool
    needs_clarification: bool
    clarification_questions: list[str]
    recommend_ready: bool
    assistant_message: str
    messages: list[InterviewMessage]

File: frontend/src/types/interview.ts
export type InterviewStatus = 'in_progress' | 'awaiting_confirmation' | 'confirmed'

export type InterviewDimension =
  | 'research_interests'
  | 'research_mode'
  | 'mentorship_style'
  | 'career_orientation'
  | 'innovation_risk'
  | 'hard_constraints'

export type HardConstraintField =
  | 'location'
  | 'weekly_commitment_days'
  | 'degree_stage'
  | 'language'
  | 'confidentiality'
  | 'graduation_arrangement'
  | 'department'
  | 'research_topic'
  | 'advisor_id'

export type HardConstraintOperator =
  | 'equals'
  | 'one_of'
  | 'excludes'
  | 'contains'
  | 'minimum'
  | 'maximum'

export interface HardConstraint {
  field: HardConstraintField
  operator: HardConstraintOperator
  value: string[]
  source_text?: string | null
}

export interface HardConstraintCapability {
  field: HardConstraintField
  label: string
  available: boolean
  evidence_record_count: number
  candidate_count: number
  evidence_coverage: number
  operators: HardConstraintOperator[]
  values: string[]
  accepts_free_text: boolean
  unavailable_reason: string | null
}

export interface HardConstraintCapabilities {
  version: 'hard-constraints-v1'
  candidate_count: number
  fields: HardConstraintCapability[]
  basis: 'published_verified_candidate_fields'
}

export interface DraftHardConstraint {
  draft_id: string
  source_text: string
  proposed_constraint: HardConstraint | null
  parsing_confidence: number
  confirmation_prompt: string
}

export interface InterviewPortrait {
  research_interests: string[]
  interest_statement: string | null
  /** 兴趣探索（活动兴趣题）选择的活动键；仅用于回显，不参与匹配 */
  activity_interests?: string[]
  research_mode: 'theory' | 'engineering' | 'mixed' | 'undecided' | null
  mentorship_style: 'high_guidance' | 'balanced' | 'autonomous' | 'undecided' | null
  career_orientation:
    | 'academic'
    | 'industry'
    | 'national_mission'
    | 'mixed'
    | 'undecided'
    | null
  innovation_risk: 'pioneering' | 'balanced' | 'mature' | 'undecided' | null
  hard_constraints: HardConstraint[] | null
  draft_hard_constraints: DraftHardConstraint[]
  unresolved_hard_constraints: string[] | null
}

// =====================================================================
// 兴趣探索（活动兴趣题 → 候选研究方向；确定性映射，GLM 不改变结果）
// =====================================================================

/** 研究场景活动选项（O*NET Interest Profiler 思路改写） */
export interface ActivityOption {
  value: string
  label: string
  description: string
}

/** 活动兴趣选择题定义 */
export interface ActivityQuestion {
  version: 'activity-interests-v1'
  prompt: string
  options: ActivityOption[]
  min_selections: number
  max_selections: number
}

/** 候选研究方向（含详细介绍与命中活动） */
export interface DirectionCandidate {
  key: string
  label: string
  description: string
  matched_activities: ActivityOption[]
  match_score: number
}

/** 候选方向建议响应 */
export interface InterestExplorationSuggestions {
  version: 'activity-interests-v1'
  basis: 'deterministic_activity_mapping'
  candidates: DirectionCandidate[]
  hint: string
}

export interface InterviewQuestionOption {
  value: string
  label: string
}

export interface InterviewQuestion {
  question_id: string
  dimension: InterviewDimension
  prompt: string
  answer_type: 'text' | 'single_choice'
  options: InterviewQuestionOption[]
  information_goal: string
}

export interface InterviewState {
  session_id: string
  status: InterviewStatus
  profile: InterviewPortrait
  profile_version: number
  current_question: InterviewQuestion | null
  completed_dimensions: InterviewDimension[]
  missing_dimensions: InterviewDimension[]
  needs_confirmation: boolean
  needs_clarification: boolean
  clarification_questions: string[]
  recommend_ready: boolean
  assistant_message?: string
  assistant_mode?: 'fixed_interview_with_optional_llm_enhancement'
  enhancement_provider?: 'glm' | null
  enhancement_status?: 'available' | 'unavailable' | 'disabled'
}

export interface InterviewEnhancementRetryResult {
  session_id: string
  text: string
  provider: 'glm'
  status: 'available'
}

export type InterviewProfilePatch = Partial<InterviewPortrait>
